persist/fs/mmap.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"fmt"
+	"os"
+
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+type mmapFileDesc struct {
+	// file is the *os.File ref to store
+	file **os.File
+	// bytes is the []byte slice ref to store the mmap'd address
+	bytes *[]byte
+	// options specifies options to use when mmaping a file
+	options mmapOptions
+}
+
+type mmapOptions struct {
+	// read is whether to make mmap bytes ref readable
+	read bool
+	// hugePages is the mmap huge pages options
+	hugePages mmapHugePagesOptions
+}
+
+type mmapHugePagesOptions struct {
+	// enabled determines if using huge pages is enabled for platforms
+	// that support it
+	enabled bool
+	// threshold determines if the size being mmap'd is greater or equal
+	// to this value to use or not use transparent huge pages if enabled
+	threshold int64
+}
+
+func mmapFiles(opener fileOpener, files map[string]mmapFileDesc) error {
+	multiErr := xerrors.NewMultiError()
+	for filePath, desc := range files {
+		fd, err := opener(filePath)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			break
+		}
+
+		b, err := mmapFile(fd, desc.options)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			break
+		}
+
+		*desc.file = fd
```

Comment:
pretty soon we'll be writing `desc->file = fd` :rofl:

Reply:
Ha, aye